src/frontends/sql/decoder_context.cc

```diff
@@ -9,61 +13,51 @@ using ir::NamedValueMap;
 using ir::Operation;
 using ir::Value;
 
-static NamedValueMap MakeNamedValueMapFromVector(absl::string_view prefix,
-                                                 std::vector<Value> vec,
-                                                 NamedValueMap map) {
-  uint64_t i = 0;
-  for (auto iter = std::make_move_iterator(vec.begin());
-       iter != std::make_move_iterator(vec.end()); ++iter) {
-    auto insert_result = map.insert({absl::StrCat(prefix, i), *iter});
-    CHECK(insert_result.second)
-        << "Collision on key " << insert_result.first->first;
-    ++i;
-  }
-  return map;
-}
-
 const Operation &DecoderContext::MakeMergeOperation(
     std::vector<Value> direct_inputs, std::vector<Value> control_inputs) {
-  // Add the direct and control inputs into the merge with predictable
-  // prefixes.
+  size_t control_start_index = direct_inputs.size();
+  CHECK(control_start_index < std::numeric_limits<int64_t>::max());
+  auto attributes =
+      ir::NamedAttributeMap({{std::string(kMergeOpControlStartIndex),
+                              ir::Attribute::Create<ir::Int64Attribute>(
+                                  static_cast<int64_t>(control_start_index))}});
+  // Combine the direct and control inputs.
+  absl::c_move(control_inputs, std::back_inserter(direct_inputs));
   return top_level_block_builder_.AddOperation(
-      merge_operator_,
-      /*attributes=*/{},
-      MakeNamedValueMapFromVector(
-          "control_input_", std::move(control_inputs),
-          MakeNamedValueMapFromVector("direct_input_", std::move(direct_inputs),
-                                      {})));
+      merge_operator_, std::move(attributes), std::move(direct_inputs));
 }
 
 const ir::Operation &DecoderContext::MakeTagTransformOperation(
     ir::Value transformed_value, absl::string_view rule_name,
     const absl::flat_hash_map<std::string, uint64_t> &preconditions) {
   // We need to reserve one space for the transformed value plus one for each
   // precondition.
-  NamedValueMap values_map;
-  values_map.reserve(preconditions.size() + 1);
+  ir::ValueList inputs;
+  // Not adding +1 to avoid potential overflow.
+  inputs.reserve(preconditions.size());
```

Comment:
I think the `+ 1` is fine, actually. If we overflow a `size_t`, it will roll right back to `0`, and `reserve`ing 0 is not a correctness concern, just a performance one. If you don't like the overflow because it is weird to reason about, I think we should just drop it and just lean on `std::vector`'s amortization argument for performance.

Reply:
Will remove reserve altogether.